chapters/functions.tex

```diff
@@ -1622,6 +1622,72 @@ \section{Declaring Inverses of Functions}\label{declaring-inverses-of-functions}
 \end{lstlisting}
 \end{example}
 
+Tools are not expected to verify the bijectiveness requirement, meaning that it is the user's responsibility to ensure that this requirement is fulfilled, and that tools can rely on the requirement as an assumption for symbolic manipulations when an inverse function is provided.
+
+There is no guarantee that a provided inverse will be used, and no rule for at which stage of symbolic processing it could be applied.
+Inlining a function means that the possibility to apply provided inverses is lost, so choosing the right variant of the function inlining alternatives can be important when inverse functions are provided.
```

Comment:
```suggestion
Inlining a function means that the possibility to apply provided inverses is lost, thus inlining should normally be performed after the inverse is potentially used. Tools should ideally do this automatically, but choosing the right variant of the function inlining alternatives can be helpful.
```

Reply:
Probably a good suggestion, but I think I just need some help here understanding the part of what tools should ideally do?

Reply:
Right (but remembering the non-inlined function seems terribly complicated).  Can this be stated more concretely as that tools making a choice in the absence of any inlining annotation should not inline more eagerly than with `LateInline = true`?

Similarly, for the case of making an explicit choice of inlining annotation, would it be too strong to recommend picking one of `Inline = false` or `LateInline = true`, and saying this before telling what tools should do?
```suggestion
Inlining a function means that the possibility to apply provided inverses is lost.
Hence, the recommended inlining annotations -- if any -- for use together with the \lstinline!inverse!-annotation is either \lstinline!Inline = false! or \lstinline!LateInline = true!.
A tool choosing when to inline in the absence of any inlining annotation should not inline more eagerly than with \lstinline!LateInline = true!.
```

Reply:
> Right (but remembering the non-inlined function seems terribly complicated). Can this be stated more concretely as that tools making a choice in the absence of any inlining annotation should not inline more eagerly than with `LateInline = true`?

I would prefer to not over-constrain implementations, and as I recall it is sometimes necessary to inline before late inlining - but I cannot remember the details at the moment.

Reply:
Sure, finding the right balance between over- and underspecifying is important, and perhaps `LateInline = true` was too restrictive.  The problem with under-specifying too much is that users end up having to set `LateInline = true` or `Inline = false` to make their inverse work as they expect across tools.  If I recall correctly, System Modeler doesn't automatically inline before `LateInline = true` when no `inverse` is present, and I am not aware of any problems due to this.  Maybe we should find a compromise to further weaken the _should not_ (which is already weaker than _must not_) to something like _is recommended not to_?

Alternatively, we could skip the entire sentence about what tools should do for now, and instead open a separate issue about the need to set either `LateInline = true` or `Inline = false` to get `inverse` functions that can be expected to work across tools.

Reply:
We also state on the line just above: "There is no guarantee that a provided inverse will be used, and no rule for at which stage of symbolic processing it could be applied." A tool deciding to inline might be one of those cases where it is not guaranteed that the inverse is used. So maybe  it is best to open a separate ticket.

Reply:
I would prefer to also remove "Hence, the recommended inlining annotations -- if any -- for use together with the \lstinline!inverse!-annotation is either \lstinline!Inline = false! or \lstinline!LateInline = true!."
So that we view it more as tool-responsibility than user-responsibility.

Reply:
I agree with the goal of making it primarily a tool responsibility, but as pointed out in #2942, one currently needs to specify inlining if one wants to be sure that the provided inverse will be applicable when needed in all tools.  The current formulation doesn't go as far as to say that an inlining alternative _should_ be specified, only which ones are reasonable to use, **if any**.

Reply:
To me that is all tool responsibility, and we don't want to tell users that they should be careful with Inline-settings if tools can and do handle it better. Saying nothing until we have resolved it in #2942 seems like the best action.

Reply:
I think the best may be to remove the lines 1628-9 (Inlining a function means that the possibility to apply provided inverses is lost.....) for now. It may be decided in #2942 that the tool should not inline them if there is the inverse annotation.

Reply:
I consider #2942 a separate question.  These two sentences are advice for humans.  Would it be better if it was formulated as a recommendation against combining with `Inline = true` or `InlineAfterIndexReduction = true`?

Reply:
I also consider it as advice to humans, and I don't want such advice for working around limitations in tools if there's no need for that. Especially not, if we will later improve the specification to avoid it altogether.

Reply:
I don't think the need to discuss what humans should or shouldn't do can be avoided entirely as long as the possibility to add explicit inlining annotations exist – even if tools are required to make something intelligent when no inlining annotations are given, it would be a bad idea to add an explicit `Inline = true`.

Reply:
I disagree. If tools can make an intelligent choice (as they can in this case), we shouldn't warn users and persuade them to go for the lowest common denominator.
We want to encourage tools (and users) to improve - and the best way to do that is not add such restrictions; not to provide work-arounds for previous or existing limitation.

Reply:
I kind of disagree to disagree.  Neither do I want to persuade users to go for a lowest common denominator by getting in the way of tools that are supposed to do the right thing by themselves.  Further, yes, we should encourage tools to improve (but not sure what you mean with users here).

To understand your conclusion, however, I need to understand better what you mean by _previous or existing limitation_.  Maybe I'm lost in this discussion because I still think of the various inlining annotations per se as a feature – not a limitation – of the language?

Reply:
I view both inverse and inlining annotations as features that takes care to implement in tools, including their interactions - and I know that we previously had limitations in Dymola in dealing with them in a good way.
Thus I prefer to not mention this in this PR, but discuss future possibilities in #2942

Reply:
But #2942 is about what tools should be expected to take care of automatically.  This is about stopping users from getting in the way of their tools by adding the most blatantly incompatible combinations of annotations.

To be concrete: Can anyone give any examples of useful combinations of `inverse` and either `Inline = true` or `InlineAfterIndexReduction = true`?  If not, why shouldn't we let the readers know that these combinations don't make much sense?

Then, once we figure out what to expect from automatic logic in tools in #2942, we should be able make even stronger recommendations for the users, telling them to avoid combining `inverse` with any of the inlining annotations.

Reply:
> But #2942 is about what tools should be expected to take care of automatically. This is about stopping users from getting in the way of their tools by adding the most blatantly incompatible combinations of annotations.
> 
> To be concrete: Can anyone give any examples of useful combinations of `inverse` and either `Inline = true` or `InlineAfterIndexReduction = true`? If not, why shouldn't we let the readers know that these combinations don't make much sense?

Ok, I'm convinced.

Reply:
Good, resolving.